src/Introspection.hack

```diff
@@ -0,0 +1,289 @@
+namespace Slack\GraphQL\Introspection;
```

Comment:
This does strike me as verbose, but I'm not sure that there's a simpler way to paper-over the difference between Input and Output types. For example, we want to be able to use both input and output enum types, scalar types, etc, interchangeably for the purpose of introspection (input and output enums are not two separate types). This lets us paper over that distinction by forcing both input and output enum types to implement `IntrospectableEnum`, presumably returning the same name and description for enums generated from the same Hack type.

Is there some other reason which I'm missing to use these interfaces?

Reply:
We could have potentially made __Type, __Field etc interfaces and then had the generated types implement those. I think the issue with that though is then those fields will conflict with any fields defined in those objects. This way, none of the introspection fields are available on the generated types (aside from within the Query itself).

Reply:
> we want to be able to use both input and output enum types, scalar types, etc, interchangeably for the purpose of introspection (input and output enums are not two separate types)

We could also make an (arbitrary) choice that either the input or the output type will implement the introspection interface, for the types that have both. It shouldn't really matter which; and it might save us some duplicate code, e.g. we won't have to implement things like getKind() on both.